### ISSUE TITLE
[Serve] ngram_ple_expand_columns writes the within-lane offset as segment_begin, not the segment's first column

### ISSUE THREAD

**Issue by flaviusburca** [invergent-ai/surogate#112]
**Severity:** high &nbsp;·&nbsp; **Location:** [`csrc/src/serve/ops/ngram_ple/ngram_ple.cu:314`](https://github.com/invergent-ai/surogate/blob/7f12387624e04c498dbd8df879eaeff13dea0b25/csrc/src/serve/ops/ngram_ple/ngram_ple.cu#L314) &nbsp;·&nbsp; **Audited at:** `7f123876`

### Defect
`expand_columns_kernel` stores the column's offset *inside* its lane (`offset = i - lane*width`) into `segment_begin`. Every consumer of `segment_begin` treats it as the ABSOLUTE column index at which the segment starts: `predecessor()` computes `source = t - k` from the absolute column `t = blockIdx.x` and tests `source >= begin`, with the comment "1 = the column just before the segment"; `conv_input()` does the same with `source = t - back`. The public contract in api/ops/ngram_ple.h:68 is explicit: segment_begin is "the column index where the column's sequence segment starts in this call". The two other stagings agree with that reading — `decode_columns` fills it with `fill_i32_positions(...,0)` which writes the ramp `begin[t]=t` (absolute), and `single_segment_columns` fills it with the constant 0 (absolute). Only the speculative/verify staging writes a lane-relative value, so for lane b the kernels compare an absolute `source` against a relative `begin`, shifting the segment boundary by `lane*width` columns. Line 314 should be `begin[i] = lane * width;` (lines 313 and 315 are already correct).

### Failure scenario
Flash-Next (qwen4_exp, PLE layer 1) speculative verify round with width=4 (K=3) and 2 lanes, so columns 0..3 are request A and 4..7 are request B; `ngram_ple_expand_columns` writes segment_begin = [0,1,2,3,0,1,2,3]. (a) Cross-request context: lane 1's first column is absolute t=4 with begin=0; `predecessor(t=4,k=1)` gives source=3 and `3 >= 0` is true, so it returns `ids[3]` — request A's fourth token — and the bigram hash `ctx[0]*m0 ^ ctx[1]*m1` selects a PLE table row derived from another request's token; with ngram=3 the trigram head also reads `ids[2]`. (b) Lost own context even at batch=1: lane 0's column t=3 has begin=3, so `source = 3-1 = 2` fails `2 >= 3` and the kernel falls to the slot history instead of returning `ids[2]`, i.e. the round's own preceding token is never seen. (c) The same `begin` drives `conv_input` with `back` up to history=9, so for lane 1 every `back <= 4` reads request A's normalised columns and every `back > 4` indexes the slot conv_state at `history-(back-4)` instead of `history-back` — a 4-tap shift on every column of every lane except lane 0. `ple_conv_state_update_kernel` then persists that corrupted window into the slot, so the damage outlives the round.

### Evidence
````cpp
ngram_ple.cu:309-316
    const int i = static_cast<int>(blockIdx.x * blockDim.x + threadIdx.x);
    if (i >= columns) { return; }
    const int lane   = i / width;
    const int offset = i - lane * width;
    slots_out[i]     = slots_in[lane];
    begin[i]         = offset;                       // <-- lane-relative
    last[i]          = offset == width - 1 ? 1 : 0;

consumer, ngram_ple.cu:39-46
__device__ __forceinline__ int predecessor(const int* ids, const int* history, int t, int k,
                                           int begin, int slot, int history_len, int slots) {
    const int source = t - k;
    if (source >= begin) { return ids[source]; }
    const int back = begin - source; // 1 = the column just before the segment

contract, csrc/src/serve/api/ops/ngram_ple.h:67-68
 * Per-column metadata for one call: `ids` I32 [T] token ids; `segment_begin` I32 [T] the
 * column index where the column's sequence segment starts in this call;
````

### Suggested fix
Change line 314 to `begin[i] = lane * width;` and fix the contradicting comment block in api/ops/ngram_ple.h:126 (`segment_begin[b*width + j] = j`) to `segment_begin[b*width + j] = b*width`. A regression test should compare a width-W batch-B verify round against B separate width-W single-segment calls; today they disagree on every column except column 0 of lane 0.

### Verification
Found in an audit of `csrc/src/serve`, then upheld by two independent verifiers: one tried to refute it, the other traced a concrete path to it.
Corrections the verifiers made to the original report:

> The defect is real and the claim is right on the essentials. Two corrections to the details:
> 1. The doc comment for ngram_ple_expand_columns at csrc/src/serve/api/ops/ngram_ple.h:124-128 has the same error: it says `segment_begin[b*width + j] = j`. It contradicts both the NgramPleColumns contract and its own prose ("The columns of one lane are one segment"), so the fix must change the header as well as ngram_ple.cu:314.
> 2. In part (c), for lane b>0 at column j, only taps with j < back <= b*width read the previous lane's columns; taps with back <= j read the correct columns of their own lane. Taps with back > b*width read conv_state at history-(back-b*width) instead of the value the correct begin would give.
> 
> At batch 1 (lane 0), every column j >= 1 takes its n-gram predecessors and conv taps from the pre-round slot state instead of from the round's own earlier columns. The history the last column persists is also wrong. So Flash-Next MTP verify logits are wrong whenever width > 1, not only when several lanes are batched.

> The defect is real. One detail in (c) is off: for lane b at offset j, when `back > b*width`, the buggy kernel reads `conv_state[history-(back-b*width)]`. The correct read is `conv_state[history-(back-j)]` (or `normalized[t-back]` when `back <= j`), not `history-back`. The header comment at api/ops/ngram_ple.h:126 (`segment_begin[b*width + j] = j`) documents the same wrong value and needs fixing along with ngram_ple.cu:314 (`begin[i] = lane * width;`). The effect is silently wrong PLE embeddings and convolution output on every column after the first in every wide verify round (width>1), including batch=1. The corrupted history and conv_state then persist to the slot. There is no crash or out-of-bounds access, since all indices stay within ids/history/conv_state.

_Line numbers quoted by the verifiers may refer to a later tree; the permalinks above are pinned to `7f123876`._

**Comment by flaviusburca** [comment]
Fixed in commit `ff2b0b52e606c41402bf6aac4a47927a6dad11a7`. The fix is committed locally and validated; publication to the remote repository is pending.

PLE speculative expansion now records each lane's first column as its segment start.

Validation: Batched and independent-sequence outputs, token histories, and convolution histories match at widths 1/4/8, including subsequent rounds and EOS. CUDA memcheck reports no errors.
